File: app/src/main/java/com/worker/natrobotcontroller/fragments/JoystickFragment.java
package com.worker.natrobotcontroller.fragments;

import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.support.v7.widget.AppCompatSpinner;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemSelectedListener;
import android.widget.SeekBar;
import android.widget.SeekBar.OnSeekBarChangeListener;
import android.widget.TextView;

import com.h6ah4i.android.widget.verticalseekbar.VerticalSeekBar;
import com.worker.natrobotcontroller.R;
import com.worker.natrobotcontroller.R.id;
import com.worker.natrobotcontroller.activities.MainActivity;

import org.jetbrains.anko.ToastsKt;

import java.io.BufferedReader;
import java.io.InputStreamReader;

import io.github.controlwear.virtual.joystick.android.JoystickView;
import io.github.controlwear.virtual.joystick.android.JoystickView.OnMoveListener;
import kotlin.text.Charsets;

public final class JoystickFragment extends Fragment {

    public String mode = "";
    public int speed = 200;
    public int frontDistance;
    public int backDistance;
    public int currentAngle;
    public int desiredAngle;
    public boolean reading = true;
    TextView info;
    private SharedPreferences preferences;
    private View v;
    private MainActivity activity;

    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        activity = (MainActivity) getActivity();
        v = inflater.inflate(R.layout.controller, container, false);
        info = v.findViewById(id.car_info);
        v.findViewById(id.fowardBtn).setOnClickListener(new OnClickListener() {
            public final void onClick(View it) {
                sendCommand("_0;s" + speed + ';');
            }
        });
        v.findViewById(id.backBtn).setOnClickListener(new OnClickListener() {
            public final void onClick(View it) {
                sendCommand("_180;s" + speed + ';');
            }
        });
        v.findViewById(id.leftBtn).setOnClickListener(new OnClickListener() {
            public final void onClick(View it) {
                sendCommand("_90;s" + speed + ';');
            }
        });
        v.findViewById(id.rightBtn).setOnClickListener(new OnClickListener() {
            public final void onClick(View it) {
                sendCommand("_-90;s" + speed + ';');
            }
        });
        v.findViewById(id.stopBtn).setOnClickListener(new OnClickListener() {
            public final void onClick(View it) {
                sendCommand("p10;");
            }
        });
        ((VerticalSeekBar) v.findViewById(id.speedBar)).setProgress(this.speed);
        ((VerticalSeekBar) v.findViewById(id.speedBar)).setOnSeekBarChangeListener(new OnSeekBarChangeListener() {
            public void onProgressChanged(SeekBar p0, int progress, boolean fromUser) {
                speed = progress;
                sendCommand("" + 's' + speed + ';');
            }

            public void onStartTrackingTouch(SeekBar p0) {
            }

            public void onStopTrackingTouch(SeekBar p0) {
            }
        });
        ((JoystickView) v.findViewById(id.joystickR)).setOnMoveListener(new OnMoveListener() {
            public final void onMove(int angle, int strength) {
                int pwm = strength * 255 / 100;
                String cmd = "" + '_' + (angle - 90) + ";s" + pwm + ';';
                sendCommand(cmd);
            }
        }, 200);
        v.findViewById(id.reset_gyro).setOnClickListener(new View.OnClickListener() {
            public final void onClick(View it) {
                sendCommand("g20;");
            }
        });

        ((AppCompatSpinner) v.findViewById(id.controller_mode)).setOnItemSelectedListener(new OnItemSelectedListener() {
            public void onNothingSelected(AdapterView p0) {
            }

            public void onItemSelected(AdapterView p0, View p1, int p2, long p3) {
                String mode = ((AppCompatSpinner) v.findViewById(id.controller_mode)).getSelectedItem().toString();
                selectMode(mode);
                getPreferences().edit().putString("control_mode", mode).apply();
            }
        });
        this.selectMode(this.mode);
        return v;

    }

    public final void startReading() {
        new ReadingTask().execute();
    }

    public final void selectMode(String mode) {
        switch (mode) {
            case "Game pad":
                if (mode.equals("Game pad")) {
                    v.findViewById(id.keyboard_layout).setVisibility(View.VISIBLE);
                    v.findViewById(id.joystickR).setVisibility(View.INVISIBLE);
                }
                break;
            case "Joystick":
                if (mode.equals("Joystick")) {
                    v.findViewById(id.keyboard_layout).setVisibility(View.INVISIBLE);
                    v.findViewById(id.joystickR).setVisibility(View.VISIBLE);
                }
        }

    }

    public final int readNumber(String s) {
        return Integer.parseInt(s.substring((byte) 1, s.length()));
    }

    public void onResume() {
        super.onResume();
        this.reloadSetting();
        this.reading = true;
        this.startReading();
    }

    public void onPause() {
        super.onPause();
        this.reading = false;
    }

    public final void reloadSetting() {
        this.mode = this.getPreferences().getString("control_mode", "Joystick");
    }

    public final void sendCommand(String command) {
        try {
            activity.socket.getOutputStream().write(command.getBytes(Charsets.UTF_8));
        } catch (Exception e) {
            e.printStackTrace();
            v.<TextView>findViewById(id.car_info).setText("Not connected");
        }
    }

    public final void slimSendCommand(String command) {
        try {
            activity.socket.getOutputStream().write(command.getBytes(Charsets.UTF_8));
        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    public final void trigger(int direction) {
        if (direction == -1) {
            this.slimSendCommand("p10;");
        } else {
            int tmp = this.currentAngle + direction * 90;
            while (tmp < 0) {
                tmp += 360;
            }
            while (tmp > 360) {
                tmp -= 360;
            }
            if (tmp < 45) {
                tmp = 0;
            } else if (tmp < 135) {
                tmp = 90;
            } else if (tmp < 225) {
                tmp = 180;
            } else if (tmp < 315) {
                tmp = 270;
            }
            this.slimSendCommand("_" + tmp + ";s250;");
        }

    }

    public SharedPreferences getPreferences() {
        if (preferences == null)
            preferences = PreferenceManager.getDefaultSharedPreferences(getActivity());
        return preferences;
    }

    class ReadingTask extends AsyncTask<Void, Void, Void> {
        @Override
        protected Void doInBackground(Void... voids) {
            while (reading) {
                try {
                    String line = new BufferedReader(new InputStreamReader(activity.socket.getInputStream())).readLine().trim();
                    if (!line.isEmpty()) {
                        for (String cmd : line.split(";")) {
                            if (!cmd.isEmpty()) {
                                switch (cmd.charAt(0)) {
                                    case 'b':
                                        backDistance = readNumber(cmd);
                                        break;
                                    case 'c':
                                        currentAngle = readNumber(cmd);
                                        break;
                                    case 'd':
                                        desiredAngle = readNumber(cmd);
                                    case 'e':
                                    default:
                                        break;
                                    case 'f':
                                        frontDistance = readNumber(cmd);
                                }
                            }
                        }
                    }
                    getActivity().runOnUiThread(new Runnable() {
                        public final void run() {
                            info.setText(String.format("Front: %d cm\nBack: %d\nCurrent angle: %d\nDesired angle: %d", frontDistance, backDistance, currentAngle, desiredAngle));
                        }
                    });
                } catch (Exception e) {
                    if (e instanceof NullPointerException)
                        try {
                            Thread.sleep(1000);
                        } catch (InterruptedException e1) {
                            e1.printStackTrace();
                        }
                    else
                        e.printStackTrace();
                }
            }
            return null;
        }
    }

}

File: app/src/main/java/com/worker/natrobotcontroller/fragments/JoystickFragment.kt
package com.worker.natrobotcontroller.fragments

import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v4.app.Fragment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.SeekBar
import com.worker.natrobotcontroller.R
import com.worker.natrobotcontroller.activities.MainActivity
import kotlinx.android.synthetic.main.controller.view.*
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.toast
import java.io.IOException

/**
 * Created by hataketsu on 11/12/17.
 */
class JoystickFragment : Fragment() {
    var mode = ""
    var speed = 200
    var frontDistance = 0
    var backDistance = 0
    var currentAngle = 0
    var desiredAngle = 0
    lateinit var connect: ConnectFragment
    val preferences by lazy {
        PreferenceManager.getDefaultSharedPreferences(context)
    }
    var reading = true

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val _connect = (activity as MainActivity).connect
        this.connect = _connect!!

        val v = inflater?.inflate(R.layout.controller, container, false)
        v!!.fowardBtn.setOnClickListener { sendCommand(connect, "_0;s$speed;") }
        v.backBtn.setOnClickListener { sendCommand(connect, "_180;s$speed;") }
        v.leftBtn.setOnClickListener { sendCommand(connect, "_90;s$speed;") }
        v.rightBtn.setOnClickListener { sendCommand(connect, "_-90;s$speed;") }
        v.stopBtn.setOnClickListener { sendCommand(connect, "p10;") }
        v.speedBar.progress = speed
        v.speedBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(p0: SeekBar?, progress: Int, fromUser: Boolean) {
                speed = progress
                sendCommand(connect, "s$speed;")
            }

            override fun onStartTrackingTouch(p0: SeekBar?) {}

            override fun onStopTrackingTouch(p0: SeekBar?) {}

        })
        v.joystickR.setOnMoveListener({ angle, strength ->
            val pwm = strength * 255 / 100
            val cmd = "_${angle - 90};s$pwm;"
            sendCommand(connect, cmd)
        }, 200)
        v.reset_gyro.setOnClickListener {
            sendCommand(connect, "g20;")
        }


        v.controller_mode.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onNothingSelected(p0: AdapterView<*>?) {}

            override fun onItemSelected(p0: AdapterView<*>?, p1: View?, p2: Int, p3: Long) {
                val mode = v.controller_mode.selectedItem.toString()
                activity.toast(mode)
                selectMode(mode)
                preferences.edit().putString("control_mode", mode).apply()
            }

        }
        selectMode(mode)

        return v
    }

    private fun startReading() {
        activity.doAsync {
            while (reading) {
                try {
                    if (connect.socket?.inputStream != null) {
                        if (connect.socket?.isConnected!!) {
                            val line=connect.socket?.inputStream?.bufferedReader()?.readLine()?.trim()
                            Log.d("BLUETOOTHx", "read" + line)
                            if (!line.isNullOrEmpty()) {
                                line!!.split(";").forEach { cmd ->
                                    if (!cmd.isEmpty())
                                        when (cmd[0]) {
                                            'f' -> frontDistance = readNumber(cmd)
                                            'b' -> backDistance = readNumber(cmd)
                                            'c' -> currentAngle = readNumber(cmd)
                                            'd' -> desiredAngle = readNumber(cmd)
                                        }
                                }

                            }
                            activity.runOnUiThread {
                                view?.car_info?.setText("Front: $frontDistance cm\n" +
                                        "Back: $backDistance\n" +
                                        "Current angle: $currentAngle\n" +
                                        "Desired angle: $desiredAngle")
                            }
                        }
                    } else
                        try {
                            Thread.sleep(1000)
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
    }

    private fun selectMode(mode: String) {
        when (mode) {
            "Joystick" -> {
                view?.keyboard_layout?.visibility = View.INVISIBLE
                view?.joystickR?.visibility = View.VISIBLE
            }
            "Game pad" -> {
                view?.keyboard_layout?.visibility = View.VISIBLE
                view?.joystickR?.visibility = View.INVISIBLE
            }
        }
    }

    private fun readNumber(s: String): Int {
        return s.substring(1, s.length).toInt()
    }

    override fun onResume() {
        super.onResume()
        reloadSetting()
        reading = true
        startReading()
    }

    override fun onPause() {
        super.onPause()
        reading = false
    }

    private fun reloadSetting() {
        mode = preferences.getString("control_mode", "Joystick")
    }

    private fun sendCommand(connect: ConnectFragment?, command: String) {
        connect?.log("Try to send $command")
        if (connect?.socket == null) {
            view?.car_info?.setText("Not connected")
        } else
            try {
                connect.socket?.outputStream?.write(command.toByteArray())
            } catch (e: IOException) {
                e.printStackTrace()
                view?.car_info?.setText("Not connected")
            }
    }

    private fun slimSendCommand(connect: ConnectFragment?, command: String) {
        connect?.log("Try to send $command")
        if (connect?.socket != null)
            try {
                connect.socket?.outputStream?.write(command.toByteArray())
            } catch (e: IOException) {
                e.printStackTrace()
            }
    }

    fun trigger(direction: Int) {
        var tmp = currentAngle + direction * 90
        while (tmp < 0) tmp += 360
        while (tmp > 360) tmp -= 360
        if (tmp < 45)
            tmp = 0
        else if (tmp < 45 + 90)
            tmp = 90
        else if (tmp < 45 + 90 + 90)
            tmp = 90 + 90
        else if (tmp < 45 + 90 + 90 + 90)
            tmp = 90 + 90 + 90
        slimSendCommand(connect, "_$tmp;s250;")
    }


}